server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java

```diff
@@ -296,13 +272,12 @@ private ClusterState applyRequest(ClusterState currentState, PutMappingClusterSt
                 final Index index = indexMetaData.getIndex();
                 final MapperService mapperService = indexMapperServices.get(index);
 
-                String typeForUpdate = mapperService.getTypeForUpdate(mappingType, mappingUpdateSource);
                 CompressedXContent existingSource = null;
-                DocumentMapper existingMapper = mapperService.documentMapper(typeForUpdate);
+                DocumentMapper existingMapper = mapperService.documentMapper();
                 if (existingMapper != null) {
                     existingSource = existingMapper.mappingSource();
                 }
-                DocumentMapper mergedMapper = mapperService.merge(typeForUpdate, mappingUpdateSource, MergeReason.MAPPING_UPDATE);
+                DocumentMapper mergedMapper = mapperService.merge(request.type(), mappingUpdateSource, MergeReason.MAPPING_UPDATE);
```

Comment:
Here we no longer use the resolved type `typeForUpdate`. This type resolution logic allowed for 'put mapping' to work in the following scenario:
- An index has a pre-existing type like `my_type`.
- The user adds new mappings using a typeless put mapping call (which uses the `_doc` type under the hood).

We still need to support this case, because there could be indices carried over from 7.x with a custom type name. I think that CI didn't catch the issue because the relevant 'mix typeful and typeless' tests were removed in https://github.com/elastic/elasticsearch/pull/41676. Perhaps we could restore some of those tests, and make sure to preserve the necessary logic here.

Reply:
Type names will be removed entirely for 8x though, so I don't think this issue will arise?  The point of this PR is to remove the type resolution checks because we know that an index can only have a single type, so it doesn't matter what it's called.  It's a staging commit that means we can remove type information from things like put-mapping or create-index without having to go and change all our tests to use a `_doc` type, only to then remove the types completely in a follow-up commit.

Reply:
Sorry @romseygeek, I should have done some tests before writing that comment! The scenario I described actually seems to work, because we completely ignore the type on the request. I don't see a bug, and it sounds like we're on the same page with this PR.

I do wonder if could use more tests around the scenario I described, I'll follow-up on #41676 to discuss.